File: platform/core/PolyPodCoreAndroid/core/src/main/kotlin/coop/polypoly/core/FeatureCategory.kt
package coop.polypoly.core

import org.msgpack.value.Value

enum class FeatureCategoryId {
    yourData,
    knowHow,
    tools,
    developer;

    companion object {
        fun from(value: Value): FeatureCategoryId {
            return FeatureCategoryId.valueOf(value.toString())
        }
    }
}

data class FeatureCategory(
    val id: FeatureCategoryId,
    val name: String,
    val features: List<Feature>
) {
    companion object {
        fun from(value: Value): FeatureCategory {
            val map = value.asMapValue().map()

            return FeatureCategory(
                FeatureCategoryId.from(map.getValue("id")),
                map.getValue("name").toString(),
                map.getValue("features").asArrayValue().map(Feature::from)
            )
        }
    }
}

data class Feature(
    val path: String,
    val id: String,
    val name: String,
    val author: String?,
    val version: String?,
    val description: String?,
    val thumbnail: String?,
    val primaryColor: Int,
    val thumbnailColor: Int,
    val borderColor: Int,
    val tileTextColor: Int,
    val links: Map<String, String>
) {
    companion object {
        fun from(value: Value): Feature {
            val map = value.asMapValue().map()

            return Feature(
                map.getString("path"),
                id = map.getString("id"),
                name = map.getString("name"),
                author = map.getOptionalString("author"),
                version = map.getOptionalString("version"),
                description = map.getOptionalString("description"),
                thumbnail = map.getOptionalString("thumbnail"),
                thumbnailColor = map.getColor("thumbnailColor"),
                primaryColor = map.getColor("primaryColor"),
                borderColor = map.getColor("borderColor"),
                tileTextColor = map.getColor("tileTextColor"),
                links = map.getValue("links").asMapValue().map().map {
                    it.key.asStringValue().asString() to
                        it.value.asStringValue().asString()
                }.toMap()
            )
        }
    }

    fun findUrl(target: String): String? = when (target) {
        in links.keys -> links[target]
        in links.values -> target
        else -> null
    }
}

File: platform/core/PolyPodCoreSwift/Sources/PolyPodCoreSwift/FeatureCategory.swift
import Foundation
import MessagePack

public enum FeatureCategoryId: String, Encodable {
    case yourData
    case knowHow
    case tools
    case developer
}

public struct FeatureCategory {
    public let id: FeatureCategoryId
    public let name: String
    public let features: [Feature]
}

public struct Feature {
    public init(
        path: URL,
        id: String,
        name: String,
        author: String?,
        description: String?,
        primaryColor: String,
        thumbnailColor: String,
        thumbnail: URL?,
        borderColor: String,
        tileTextColor: String,
        links: [String : String]
    ) {
        self.path = path
        self.id = id
        self.name = name
        self.author = author
        self.description = description
        self.primaryColor = primaryColor
        self.thumbnailColor = thumbnailColor
        self.thumbnail = thumbnail
        self.borderColor = borderColor
        self.tileTextColor = tileTextColor
        self.links = links
    }
    
    public let path: URL
    public let id: String
    public let name: String
    public let author: String?
    public let description: String?
    public let primaryColor: String
    public let thumbnailColor: String
    public let thumbnail: URL?
    public let borderColor: String
    public let tileTextColor: String
    public let links: [String: String]
}

extension Feature {
    public func findUrl(target: String) -> String? {
        if let url = links[target] {
            return url
        }
        if links.values.contains(target) {
            return target
        }
        return nil
    }
}

extension FeatureCategory: MessagePackDecodable {
    public init(from value: MessagePackValue) throws {
        let dictionary: [MessagePackValue: MessagePackValue] = try value.getDictionary()
        self.id = try FeatureCategoryId.init(from: dictionary.get("id"))
        self.name = try dictionary.get("name").getString()
        self.features = try [Feature].init(from: dictionary.get("features"))
    }
}

extension FeatureCategoryId: MessagePackDecodable {
    public init(from value: MessagePackValue) throws {
        guard let id = FeatureCategoryId(rawValue: try value.getString()) else {
            throw DecodingError.unknownFeatureCategoryId(info: "\(value)")
        }
        self = id
    }
}

extension Feature: MessagePackDecodable {
    public init(from value: MessagePackValue) throws {
        let dictionary: CoreResponseObject = try value.getDictionary()
        
        let links = try dictionary["links"]?
            .getDictionary().reduce(into: [String: String]()) { partialResult, keyValue in
                let key: String = try keyValue.key.getString()
                let value: String = try keyValue.value.getString()
                partialResult[key] = value
            }
        path = try URL(fileURLWithPath: dictionary.get("path").getString())
        id = try dictionary.get("id").getString()
        name = try dictionary.get("name").getString()
        author = try dictionary["author"]?.getString()
        description = try dictionary["description"]?.getString()
        primaryColor = try dictionary.get("primaryColor").getString()
        thumbnailColor = try dictionary.get("thumbnailColor").getString()
        thumbnail = try dictionary["thumbnail"]?
            .getString()
            .map(URL.init(fileURLWithPath:))
        borderColor = try dictionary.get("borderColor").getString()
        tileTextColor = try dictionary.get("tileTextColor").getString()
        self.links = links ?? [:]
    }
}
